plugin/storage/cassandra/dependencystore/storage.go

```diff
@@ -26,60 +28,96 @@ import (
 	casMetrics "github.com/jaegertracing/jaeger/pkg/cassandra/metrics"
 )
 
+// IndexMode determines how the dependency data is indexed.
+type IndexMode int
+
 const (
-	depsInsertStmt = "INSERT INTO dependencies(ts, ts_index, dependencies) VALUES (?, ?, ?)"
-	depsSelectStmt = "SELECT ts, dependencies FROM dependencies WHERE ts_index >= ? AND ts_index < ?"
+	// SASIEnabled is used when the dependency table is SASI indexed.
+	SASIEnabled IndexMode = iota
+
+	// SASIDisabled is used when the dependency table is NOT SASI indexed.
+	SASIDisabled
+
+	depsInsertStmtSASI = "INSERT INTO dependencies(ts, ts_index, dependencies) VALUES (?, ?, ?)"
+	depsInsertStmt     = "INSERT INTO dependenciesv2(ts, date_bucket, dependencies) VALUES (?, ?, ?)"
+	depsSelectStmtSASI = "SELECT ts, dependencies FROM dependencies WHERE ts_index >= ? AND ts_index < ?"
+	depsSelectFmt      = "SELECT ts, dependencies FROM dependenciesv2 WHERE date_bucket IN (%s) AND ts >= ? AND ts < ?"
+	dateFmt            = "20060102"
+	day                = 24 * time.Hour
 )
 
 // DependencyStore handles all queries and insertions to Cassandra dependencies
 type DependencyStore struct {
 	session                  cassandra.Session
 	dependenciesTableMetrics *casMetrics.Table
 	logger                   *zap.Logger
+	indexMode                IndexMode
 }
 
 // NewDependencyStore returns a DependencyStore
 func NewDependencyStore(
 	session cassandra.Session,
 	metricsFactory metrics.Factory,
 	logger *zap.Logger,
+	indexMode IndexMode,
 ) *DependencyStore {
 	return &DependencyStore{
 		session:                  session,
 		dependenciesTableMetrics: casMetrics.NewTable(metricsFactory, "dependencies"),
 		logger:                   logger,
+		indexMode:                indexMode,
 	}
 }
 
 // WriteDependencies implements dependencystore.Writer#WriteDependencies.
 func (s *DependencyStore) WriteDependencies(ts time.Time, dependencies []model.DependencyLink) error {
 	deps := make([]Dependency, len(dependencies))
 	for i, d := range dependencies {
-		deps[i] = Dependency{
+		dep := Dependency{
 			Parent:    d.Parent,
 			Child:     d.Child,
 			CallCount: int64(d.CallCount),
 		}
+		if s.indexMode == SASIEnabled {
+			dep.Source = string(d.Source)
+		}
+		deps[i] = dep
+	}
+
+	var query cassandra.Query
+	switch s.indexMode {
+	case SASIDisabled:
+		query = s.session.Query(depsInsertStmt, ts, ts.Format(dateFmt), deps)
+	case SASIEnabled:
+		query = s.session.Query(depsInsertStmtSASI, ts, ts, deps)
 	}
-	query := s.session.Query(depsInsertStmt, ts, ts, deps)
 	return s.dependenciesTableMetrics.Exec(query, s.logger)
 }
 
 // GetDependencies returns all interservice dependencies
 func (s *DependencyStore) GetDependencies(endTs time.Time, lookback time.Duration) ([]model.DependencyLink, error) {
-	query := s.session.Query(depsSelectStmt, endTs.Add(-1*lookback), endTs)
+	startTs := endTs.Add(-1 * lookback)
+	var query cassandra.Query
+	switch s.indexMode {
```

Comment:
Would `GetDependencies` be simplified by omitting this switch? Instead we might read from SASIDisabled first, and if that fails we can read from SASIEnabled?
When migrating, this would mean that users don't need to redeploy query after moving over the dependency table

Reply:
I'm not a fan of adding the extra roundtrip time. Additionally, if the user has opted in to migrating to the new table, redeploying the query service will not be the most strenuous thing.